### PR TITLE
ci: test on 3.12 where appropriate

### PR DESCRIPTION
**Pull request by emmettbutler**
This change updates many test suites to include Python 3.12 in their set of tested runtimes. The ones that haven't been updated in this way are those that appear to require code changes for 3.12 support to work. This change matters because 3.12 is among the officially supported runtimes by this library.

Most of the line changes in this diff are in generated riot lockfiles and can be ignored.

The removal of the snapshot assertion in the test harness made some asynchronous tests pass without breaking any others.

## Checklist

- [x] The PR description includes an overview of the change
- [x] The PR description articulates the motivation for the change
- [x] The change includes tests OR the PR description describes a testing strategy
- [x] The PR description notes risks associated with the change, if any
- [x] Newly-added code is easy to change
- [x] The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- [x] The change includes or references documentation updates if necessary
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Newly-added code is easy to change
- [x] Release note makes sense to a user of the library
- [x] If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Review by christophe-papazian — COMMENTED** [review]
Why removing all the `@mark_asyncio` ?

File: riotfile.py [hunk 1]
```diff
@@ -190,7 +190,7 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
         ),
         Venv(
             name="appsec_iast_tdd_propagation",
-            pys=select_pys(min_version="3.11", max_version="3.11"),
+            pys=select_pys(min_version="3.11", max_version="3.12"),
```
Comment (brettlangdon) [inline riotfile.py:193]:
```suggestion
            pys=select_pys(min_version="3.11"),
```

File: riotfile.py [hunk 2]
```diff
@@ -445,19 +445,31 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
                 "gevent": latest,
-                "pytest-asyncio": "~=0.21.1",
                 "pytest-randomly": latest,
                 "python-json-logger": "==2.0.7",
                 "pyfakefs": latest,
             },
-            pys=select_pys(min_version="3.7", max_version="3.12"),
+            venvs=[
+                Venv(
+                    pys=select_pys(min_version="3.7", max_version="3.7"),
```
Comment (brettlangdon) [inline riotfile.py:454]:
```suggestion
                    pys="3.7",
```

File: riotfile.py [hunk 3]
```diff
@@ -900,15 +918,28 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
...
+            venvs=[
+                Venv(
+                    pys=select_pys(min_version="3.8", max_version="3.11"),
+                    pkgs={
+                        "sqlalchemy": "~=1.2.18",
+                        "django": "==2.2.1",
+                    },
+                ),
+                Venv(
+                    pys=select_pys(min_version="3.12", max_version="3.12"),
```
Comment (brettlangdon) [inline riotfile.py:936]:
```suggestion
                    pys=select_pys(min_version="3.12"),
```

File: riotfile.py [hunk 4]
```diff
@@ -1163,6 +1204,11 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
                     pys="3.11",
                     pkgs={"mysql-connector-python": ["~=8.0.31", latest]},
                 ),
+                Venv(
+                    # mysql-connector-python added support for Python 3.11 in 8.0.31
+                    pys="3.12",
```
Comment (brettlangdon) [inline riotfile.py:1209]: I wonder if all of these need to be `select_pys(min_version="3.12")` since we'll be adding 3.13 soon ?

**Review by brettlangdon — COMMENTED** [review]
started a review

**Review by christophe-papazian — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Comment by emmettbutler** [comment]
@Yun-Kim could you give your opinion on the MLObs aspects of this change? Thanks!

File: riotfile.py [hunk 5]
```diff
@@ -2482,32 +2617,18 @@ def select_pys(min_version=MIN_PYTHON_VERSION, max_version=MAX_PYTHON_VERSION):
                         "langchain-cohere": latest,
                         "openai": latest,
                         "pinecone-client": latest,
-                        "botocore": latest,
-                        "cohere": latest,
-                    }
-                ),
-                Venv(
```
Comment (Yun-Kim) [inline riotfile.py:2489]: I'd prefer to keep this extra venv testing the latest release of langchain, as we've been bit hard by surprise minor releases that include breaking changes.

**Review by Yun-Kim — COMMENTED** [review]
Just one request regarding a langchain venv being removed but otherwise happy to start testing on 3.12

File: riotfile.py (hunk 5 above)
Comment (emmettbutler) [reply riotfile.py:2489]: That's actually the situation that led me to remove it, so it's working as expected. Can we add it back in a separate pull request once the breaking version is supported?

**Review by Yun-Kim — APPROVED** [review]

File: riotfile.py (hunk 5 above)
Comment (Yun-Kim) [reply riotfile.py:2489]: Will add this back in in #9902.

**Pull request merged**